File: apps/android-frontend/android/app/src/main/java/au/com/codecapers/photosphere/jsengine/FfmpegKitRunner.java
package au.com.codecapers.photosphere.jsengine;

import java.lang.reflect.Method;

//
// Android ffmpeg/ffprobe runner backed by an FFmpegKit fork (for example the 16KB-page-aligned
// com.moizhassan.ffmpeg:ffmpeg-kit-16kb fork). The com.arthenica FFmpegKit API is reached by
// reflection so this class compiles whether or not the FFmpegKit dependency is on the classpath: when
// the dependency is present the argv runs in-process; when it is absent every call returns a "not
// linked" result rather than failing the build or crashing the engine. This runner handles only
// ffprobe/ffmpeg; imagemagick stays on its own runner.
//
public class FfmpegKitRunner implements MediaToolRunner {

    //
    // Identifier for this runner, surfaced in diagnostics.
    //
    public static final String ID = "android-ffmpeg-kit-fork";

    //
    // Runner id.
    //
    @Override
    public String id() {
        return ID;
    }

    //
    // Runs an ffprobe argv in-process and returns its exit code and captured output (the ffprobe JSON).
    //
    @Override
    public ToolResult ffprobe(String[] args) {
        return runViaReflection("com.arthenica.ffmpegkit.FFprobeKit", args, "ffprobe");
    }

    //
    // Runs an ffmpeg argv in-process (e.g. the screenshot extraction) and returns its exit code.
    //
    @Override
    public ToolResult ffmpeg(String[] args) {
        return runViaReflection("com.arthenica.ffmpegkit.FFmpegKit", args, "ffmpeg");
    }

    //
    // Not handled here; the bridge routes imagemagick to the ImageMagick runner.
    //
    @Override
    public ToolResult imagemagick(String[] args) {
        return new ToolResult(-1, "imagemagick not handled by FfmpegKitRunner");
    }

    //
    // Invokes <kitClassName>.executeWithArguments(String[]) reflectively, then reads
    // session.getReturnCode().getValue() and session.getOutput(). Any reflection failure (the
    // dependency is absent) yields a "not linked" result.
    //
    private ToolResult runViaReflection(String kitClassName, String[] args, String toolName) {
        try {
            Class<?> kitClass = Class.forName(kitClassName);
            Method execute = kitClass.getMethod("executeWithArguments", String[].class);
            Object session = execute.invoke(null, (Object) args);
            if (session == null) {
                return new ToolResult(-1, toolName + ": null session");
            }

            Object returnCode = session.getClass().getMethod("getReturnCode").invoke(session);
            int exitCode = -1;
            if (returnCode != null) {
                Object value = returnCode.getClass().getMethod("getValue").invoke(returnCode);
                if (value instanceof Integer) {
                    exitCode = (Integer) value;
                }
            }

            Object output = session.getClass().getMethod("getOutput").invoke(session);
            String outputText = output instanceof String ? (String) output : "";
            return new ToolResult(exitCode, outputText);
        }
        catch (Throwable error) {
            return new ToolResult(-1, toolName + " not linked (FFmpegKit dependency not present)");
        }
    }
}

File: apps/ios-frontend/ios/App/App/JsEngine/FfmpegKitRunner.swift
import Foundation

//
// iOS ffmpeg/ffprobe runner backed by the FFmpegKit SPM fork (ffmpeg-kit-full-spm). The
// com.arthenica FFmpegKit API is identical across the surviving forks, so this code is stable even if
// the chosen fork changes. imagemagick is handled by the ImageMagick runner.
//
// The body is gated on `canImport(ffmpegkit)` so the App target still compiles before the SPM package
// is added (it then reports "not linked"). Add the package in Xcode (File > Add Package Dependencies)
// to activate it; see the iOS section of README.md.
//
#if canImport(ffmpegkit)
import ffmpegkit
#endif

struct FfmpegKitRunner: MediaToolRunner {

    //
    // Identifier for this runner, surfaced in diagnostics.
    //
    let id = "ios-ffmpeg-kit-fork"

    //
    // Runs an ffprobe argv in-process and returns its exit code and captured output (the ffprobe JSON).
    //
    func ffprobe(_ args: [String]) -> ToolResult {
        #if canImport(ffmpegkit)
        let session = FFprobeKit.execute(withArguments: args)
        let exitCode = Int(session?.getReturnCode()?.getValue() ?? -1)
        let output = session?.getOutput() ?? ""
        return ToolResult(exitCode: exitCode, output: output)
        #else
        return ToolResult(exitCode: -1, output: "ffmpegkit not linked")
        #endif
    }

    //
    // Runs an ffmpeg argv in-process (e.g. the screenshot extraction) and returns its exit code.
    //
    func ffmpeg(_ args: [String]) -> ToolResult {
        #if canImport(ffmpegkit)
        let session = FFmpegKit.execute(withArguments: args)
        let exitCode = Int(session?.getReturnCode()?.getValue() ?? -1)
        let output = session?.getOutput() ?? ""
        return ToolResult(exitCode: exitCode, output: output)
        #else
        return ToolResult(exitCode: -1, output: "ffmpegkit not linked")
        #endif
    }

    //
    // Not handled here; the bridge routes imagemagick to the ImageMagick runner.
    //
    func imagemagick(_ args: [String]) -> ToolResult {
        return ToolResult(exitCode: -1, output: "imagemagick not handled by FfmpegKitRunner")
    }
}
